extensions/amp-carousel/0.1/amp-carousel.js

```diff
@@ -20,18 +20,18 @@ import {AmpScrollableCarousel} from './scrollable-carousel';
 import {CSS} from '../../../build/amp-carousel-0.1.css';
 import {isExperimentOn} from '../../../src/experiments';
 
-class CarouselSelector {
-
-  constructor(element) {
-    if (element.getAttribute('type') == 'slides') {
-      return new AmpSlideScroll(element);
+class CarouselSelector extends AMP.BaseElement {
+  /** @override */
+  upgradeCallback() {
+    if (this.element.getAttribute('type') == 'slides') {
```

Comment:
This isn't yet fixing the bug. The bug is that this is called before any attributes are passed.

The correct solution is to return a promise from this.

Just wrapping everything in

```JS
return waitForChildPromise(this.element, () => {
  return this.childNodes.length > 0;
}).then(/* existing code in upgradeCallback */);
```

should work.